NWNXLib/Assert.cpp

```diff
@@ -0,0 +1,109 @@
+#include "Assert.hpp"
+
+#include <cstdlib>
+#include <cstring>
+
+#ifdef _WIN32
+    #include "Windows.h"
+#else
+    #include <execinfo.h>
+    #include <signal.h>
+#endif
+
+namespace NWNXLib {
+
+namespace Assert {
+
+void Fail(const char* condition, const char* file, int line, const char* message)
+{
+    char buffer[2048];
+
+    if (condition)
+    {
+        std::sprintf(buffer, "ASSERTION FAILURE\n  Summary: (%s) failed at (%s:%d)", condition, file, line);
+    }
+    else
+    {
+        std::sprintf(buffer, "ASSERTION FAILURE\n  Summary: Failed at (%s:%d)", file, line);
+    }
+
+    if (message)
+    {
+        std::strcat(buffer, "\n  Message: ");
+        std::strcat(buffer, message);
+    }
+
+#ifdef _WIN32
+    void* stackTrace[20];
+    int numCapturedFrames = CaptureStackBackTrace(0, 20, stackTrace, NULL);
+
+    if (numCapturedFrames)
+    {
+        std::strcat(buffer, "\n  Backtrace:\n");
+        for (int i = 0; i < numCapturedFrames; ++i)
+        {
+            char backtraceBuffer[32];
+            std::sprintf(backtraceBuffer, "    [0x%p]\n", stackTrace[i]);
+            std::strcat(buffer, backtraceBuffer);
+        }
+    }
+#else
+    void* stackTrace[20];
+    int numCapturedFrames = backtrace(stackTrace, 20);
+
+    if (numCapturedFrames)
+    {
+        char** resolvedFrames = backtrace_symbols(stackTrace, 20);
+        std::strcat(buffer, "\n  Backtrace:\n");
+        for (int i = 0; i < numCapturedFrames; ++i)
+        {
+            char backtraceBuffer[256];
+            std::sprintf(backtraceBuffer, "    %s\n", resolvedFrames[i]);
+            std::strcat(buffer, backtraceBuffer);
+        }
+    }
+#endif // OS_WINDOWS
+
+    std::fprintf(stderr, buffer);
```

Comment:
Even though stderr is line buffered, Windows has a habit of not actually flushing it always. `std::fflush(stderr);` here wouldn't hurt :)